Gems/Pointcloud/Code/Source/Render/PointcloudFeatureProcessor.cpp

```diff
@@ -0,0 +1,265 @@
+        uint32_t vertexCount)
+    {
+        AZ::RHI::DrawLinear drawLinear;
+        drawLinear.m_vertexCount = vertexCount;
+        drawLinear.m_vertexOffset = 0;
+        drawLinear.m_instanceCount = 1;
+        drawLinear.m_instanceOffset = 0;
+
+        AZ::RHI::DrawPacketBuilder drawPacketBuilder;
+        drawPacketBuilder.Begin(nullptr);
+        drawPacketBuilder.SetDrawArguments(drawLinear);
+        drawPacketBuilder.AddShaderResourceGroup(srg->GetRHIShaderResourceGroup());
+
+        AZ::RHI::DrawPacketBuilder::DrawRequest drawRequest;
+        drawRequest.m_listTag = drawListTag;
+        drawRequest.m_pipelineState = pipelineState->GetRHIPipelineState();
+        drawRequest.m_streamBufferViews = streamBufferViews;
+        drawPacketBuilder.AddDrawItem(drawRequest);
+        return drawPacketBuilder.End();
+    }
+
+    void PointcloudFeatureProcessor::OnRenderPipelineChanged(
+        [[maybe_unused]] AZ::RPI::RenderPipeline* renderPipeline, AZ::RPI::SceneNotification::RenderPipelineChangeType changeType)
+    {
+        if (changeType == AZ::RPI::SceneNotification::RenderPipelineChangeType::Added)
+        {
+            if (!m_meshPipelineState)
+            {
+                m_meshPipelineState = aznew AZ::RPI::PipelineStateForDraw;
+                m_meshPipelineState->Init(m_shader);
+
+                AZ::RHI::InputStreamLayoutBuilder layoutBuilder;
+                layoutBuilder.AddBuffer()
+                    ->Channel("POSITION", AZ::RHI::Format::R32G32B32_FLOAT)
+                    ->Channel("COLOR", AZ::RHI::Format::R8G8B8A8_UNORM);
+                layoutBuilder.SetTopology(AZ::RHI::PrimitiveTopology::PointList);
+                auto inputStreamLayout = layoutBuilder.End();
+
+                m_meshPipelineState->SetInputStreamLayout(inputStreamLayout);
+                m_meshPipelineState->SetOutputFromScene(GetParentScene());
+                m_meshPipelineState->Finalize();
+
+                UpdateDrawPacket();
+            }
+        }
+        else if (changeType == AZ::RPI::SceneNotification::RenderPipelineChangeType::PassChanged)
+        {
+            if (m_meshPipelineState)
+            {
+                m_meshPipelineState->SetOutputFromScene(GetParentScene());
+                m_meshPipelineState->Finalize();
+                UpdateDrawPacket();
+            }
+        }
+    }
+
+    void PointcloudFeatureProcessor::UpdateShaderConstants()
+    {
+        for (auto& [_, pcData] : m_pointcloudData)
+        {
+            if (m_drawSrg && pcData.m_needSrgUpdate)
+            {
+                pcData.m_needSrgUpdate = false;
+                AZ::Matrix4x4 orientation = AZ::Matrix4x4::CreateFromTransform(pcData.m_transform);
+                m_drawSrg->SetConstant(m_modelMatrixIndex, orientation);
+                m_drawSrg->SetConstant(m_pointSizeIndex, pcData.m_pointSize);
+                m_drawSrg->Compile();
+            }
+        }
```

Comment:
Curious about this loop... 
What happens if there's more than one point cloud active at a time?
From the looks of it, the last one in the container will override everything else.

Reply:
Oh I can see, I cannot share m_drawSrg between instances. In the code you reviewed, the last pointcloud "moved" others. I will fix it.